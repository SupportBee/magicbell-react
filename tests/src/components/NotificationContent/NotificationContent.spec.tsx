import { useNotificationFactory } from '@magicbell/react-headless';
import { act, render, RenderResult } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import Sinon from 'sinon';
import { NotificationContent } from '../../../../src';
import { sampleNotification } from '../../../factories/NotificationFactory';

describe('components', () => {
  describe('NotificationContent', () => {
    let view: RenderResult;

    describe('render', () => {
      describe('sanitization', () => {
        it('renders the content as returned from the API', async () => {
          const { result } = renderHook(() =>
            useNotificationFactory({
              ...sampleNotification,
              content: `
            <p>Hello, <a href="javascript:alert('Hacked')">click here.</a></p>
            <img src="null" onerror="alert('Hacked')">
            `,
            }),
          );

          await act(async () => {
            view = render(<NotificationContent notification={result.current} />);
          });

          expect(view.container).toMatchSnapshot();
        });
      });

      describe('relative datetime', () => {
        let clock: Sinon.SinonFakeTimers;

        beforeEach(() => {
          clock = Sinon.useFakeTimers(1615373877120);
        });

        afterEach(() => {
          clock.restore();
        });

        it('replaces the content of time tags with a relative datetime', async () => {
          const { result } = renderHook(() =>
            useNotificationFactory({
              ...sampleNotification,
              content:
                '<p>The event starts <time datetime="2021-03-11T05:33:12Z">on March 11</time>.</p>',
            }),
          );

          await act(async () => {
            view = render(<NotificationContent notification={result.current} />);
          });

          expect(view.container).toMatchSnapshot();
        });
      });
    });
  });
});
