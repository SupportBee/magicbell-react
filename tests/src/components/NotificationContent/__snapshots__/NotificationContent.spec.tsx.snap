// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components NotificationContent render relative datetime replaces the content of time tags with a relative datetime 1`] = `
.emotion-0 {
  color: inherit!important;
  cursor: inherit;
  margin: 0;
  line-height: 1.2;
  font-size: 0.865em!important;
  margin-top: 1em!important;
  word-break: break-word!important;
}

<div>
  <article
    class="emotion-0"
  >
    <p>
      The event starts 
      <time
        datetime="2021-03-11T05:33:12Z"
        timeago-id="1"
      >
        on March 11
      </time>
      .
    </p>
  </article>
</div>
`;

exports[`components NotificationContent render sanitization renders the content compiled and sanitized 1`] = `
.emotion-0 {
  color: inherit!important;
  cursor: inherit;
  margin: 0;
  line-height: 1.2;
  font-size: 0.865em!important;
  margin-top: 1em!important;
  word-break: break-word!important;
}

<div>
  <article
    class="emotion-0"
  >
    
            
    <p>
      Hello, 
      <a>
        click here.
      </a>
    </p>
    
            
    <img
      src="null"
    />
    
            
  </article>
</div>
`;
